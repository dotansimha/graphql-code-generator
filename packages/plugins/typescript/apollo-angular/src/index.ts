import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { visit, GraphQLSchema, concatAST, Kind, FragmentDefinitionNode, OperationDefinitionNode } from 'graphql';
import { RawClientSideBasePluginConfig, LoadedFragment } from '@graphql-codegen/visitor-plugin-common';
import { ApolloAngularVisitor } from './visitor';
import { extname } from 'path';
import gql from 'graphql-tag';

export interface ApolloAngularRawPluginConfig extends RawClientSideBasePluginConfig {
  /**
   * @name ngModule
   * @type string
   * @description Allows to define `ngModule` as part of the plugin's config so it's globally available.
   *
   * @example graphql.macro
   * ```yml
   * config:
   *   ngModule: ./path/to/module#MyModule
   * ```
   */
  ngModule?: string;
  /**
   * @name namedClient
   * @type string
   * @description Defined the global value of `namedClient`.
   *
   * @example graphql.macro
   * ```yml
   * config:
   *   namedClient: 'customName'
   * ```
   */
  namedClient?: string;
}

export const plugin: PluginFunction<ApolloAngularRawPluginConfig> = (schema: GraphQLSchema, documents: Types.DocumentFile[], config) => {
  const allAst = concatAST(
    documents.reduce((prev, v) => {
      return [...prev, v.content];
    }, [])
  );
  const operations = allAst.definitions.filter(d => d.kind === Kind.OPERATION_DEFINITION) as OperationDefinitionNode[];
  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(fragmentDef => ({ node: fragmentDef, name: fragmentDef.name.value, onType: fragmentDef.typeCondition.name.value, isExternal: false })),
    ...(config.externalFragments || []),
  ];

  const visitor = new ApolloAngularVisitor(schema, allFragments, operations, config) as any;
  const visitorResult = visit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: [visitor.fragments, ...visitorResult.definitions.filter(t => typeof t === 'string')].join('\n'),
  };
};

export const addToSchema = gql`
  directive @NgModule(module: String!) on OBJECT | FIELD
  directive @namedClient(name: String!) on OBJECT | FIELD
`;

export const validate: PluginValidateFn<any> = async (schema: GraphQLSchema, documents: Types.DocumentFile[], config, outputFile: string) => {
  if (extname(outputFile) !== '.ts') {
    throw new Error(`Plugin "apollo-angular" requires extension to be ".ts"!`);
  }
};

export { ApolloAngularVisitor };
