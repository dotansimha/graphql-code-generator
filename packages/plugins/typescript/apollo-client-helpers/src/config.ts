export type ApolloClientHelpersConfig = {
  /**
   * @name useTypeImports
   * @type boolean
   * @default false
   * @description Will use `import type {}` rather than `import {}` when importing only types. This gives
   * compatibility with TypeScript's "importsNotUsedAsValues": "error" option
   *
   * @example
   * ```yml
   * config:
   *   useTypeImports: true
   * ```
   */
  useTypeImports?: boolean;
  /**
   * @name requireKeyFields
   * @type boolean
   * @default false
   * @description Remove optional sign from all `keyFields` fields.
   *
   */
  requireKeyFields?: boolean;
  /**
   * @name requirePoliciesForAllTypes
   * @type boolean
   * @default false
   * @description Remove optional sign from all generated keys of the root TypePolicy.
   *
   */
  requirePoliciesForAllTypes?: boolean;
};
