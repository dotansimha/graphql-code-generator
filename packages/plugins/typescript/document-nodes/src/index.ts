import { PluginFunction, PluginValidateFn, Types } from '@graphql-codegen/plugin-helpers';
import {
  NamingConvention,
  LoadedFragment,
  RawClientSideBasePluginConfig,
} from '@graphql-codegen/visitor-plugin-common';
import { GraphQLSchema, visit, concatAST, FragmentDefinitionNode, Kind } from 'graphql';
import { TypeScriptDocumentNodesVisitor } from './visitor';

/**
 * @description This plugin generates TypeScript source (`.ts`) file from GraphQL files (`.graphql`).
 */
export interface TypeScriptDocumentNodesRawPluginConfig extends RawClientSideBasePluginConfig {
  /**
   * @default change-case-all#pascalCase
   * @description Allow you to override the naming convention of the output.
   * You can either override all namings, or specify an object with specific custom naming convention per output.
   * The format of the converter must be a valid `module#method`.
   * Allowed values for specific output are: `typeNames`, `enumValues`.
   * You can also use "keep" to keep all GraphQL names as-is.
   * Additionally you can set `transformUnderscore` to `true` if you want to override the default behavior,
   * which is to preserves underscores.
   *
   * Available case functions in `change-case-all` are `camelCase`, `capitalCase`, `constantCase`, `dotCase`, `headerCase`, `noCase`, `paramCase`, `pascalCase`, `pathCase`, `sentenceCase`, `snakeCase`, `lowerCase`, `localeLowerCase`, `lowerCaseFirst`, `spongeCase`, `titleCase`, `upperCase`, `localeUpperCase` and `upperCaseFirst`
   * [See more](https://github.com/btxtiger/change-case-all)
   *
   * @exampleMarkdown
   * ## Override All Names
   * ```yml
   * config:
   *   namingConvention: change-case-all#lowerCase
   * ```
   *
   * ## Upper-case enum values
   * ```yml
   * config:
   *   namingConvention:
   *     typeNames: change-case-all#pascalCase
   *     enumValues: change-case-all#upperCase
   * ```
   *
   * ## Keep name as-is
   * ```yml
   * config:
   *   namingConvention: keep
   * ```
   *
   * ## Remove Underscores
   * ```yml
   * config:
   *   namingConvention:
   *     typeNames: change-case-all#pascalCase
   *     transformUnderscore: true
   * ```
   */
  namingConvention?: NamingConvention;
  /**
   * @default ""
   * @description Adds prefix to the name
   *
   * @exampleMarkdown
   * ```yml
   *  documents: src/api/user-service/queries.graphql
   *  generates:
   *    src/api/user-service/queries.ts:
   *    plugins:
   *      - typescript-document-nodes
   *    config:
   *      namePrefix: 'gql'
   * ```
   */
  namePrefix?: string;
  /**
   * @default ""
   * @description Adds suffix to the name
   *
   * @exampleMarkdown
   * ```yml
   *  documents: src/api/user-service/queries.graphql
   *  generates:
   *    src/api/user-service/queries.ts:
   *    plugins:
   *      - typescript-document-nodes
   *    config:
   *      nameSuffix: 'Query'
   * ```
   */
  nameSuffix?: string;
  /**
   * @default ""
   * @description Adds prefix to the fragment variable
   */
  fragmentPrefix?: string;
  /**
   * @default ""
   * @description Adds suffix to the fragment variable
   */
  fragmentSuffix?: string;
}

export const plugin: PluginFunction<TypeScriptDocumentNodesRawPluginConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: TypeScriptDocumentNodesRawPluginConfig
) => {
  const allAst = concatAST(documents.map(v => v.document));

  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];

  const visitor = new TypeScriptDocumentNodesVisitor(schema, allFragments, config, documents);
  const visitorResult = visit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: [visitor.fragments, ...visitorResult.definitions.filter(t => typeof t === 'string')].join('\n'),
  };
};

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: any,
  outputFile: string
) => {
  if (!outputFile.endsWith('.ts')) {
    throw new Error(`Plugin "typescript-document-nodes" requires extension to be ".ts"!`);
  }
};
