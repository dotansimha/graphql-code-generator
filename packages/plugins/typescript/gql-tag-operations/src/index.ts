import { PluginFunction } from '@graphql-codegen/plugin-helpers';
import { FragmentDefinitionNode, OperationDefinitionNode } from 'graphql';
import { Source } from '@graphql-tools/utils';

export type OperationOrFragment = {
  initialName: string;
  definition: OperationDefinitionNode | FragmentDefinitionNode;
};

export type SourceWithOperations = {
  source: Source;
  operations: Array<OperationOrFragment>;
};

const documentTypePartial = `
export type DocumentType<TDocumentNode extends DocumentNode<any, any>> = TDocumentNode extends DocumentNode<
  infer TType,
  any
>
  ? TType
  : never;
`;

export const plugin: PluginFunction<{
  sourcesWithOperations: Array<SourceWithOperations>;
  useTypeImports?: boolean;
}> = (_, __, { sourcesWithOperations, useTypeImports }, _info) => {
  if (!sourcesWithOperations) {
    return '';
  }
  return [
    `import * as graphql from './graphql';\n`,
    `${
      useTypeImports ? 'import type' : 'import'
    } { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core';\n`,
    `\n`,
    ...getDocumentRegistryChunk(sourcesWithOperations),
    `\n`,
    ...getGqlOverloadChunk(sourcesWithOperations),
    `\n`,
    `export function gql(source: string): unknown;\n`,
    `export function gql(source: string) {\n`,
    `  return (documents as any)[source] ?? {};\n`,
    `}\n`,
    documentTypePartial,
  ].join(``);
};

function getDocumentRegistryChunk(sourcesWithOperations: Array<SourceWithOperations> = []) {
  const lines: Array<string> = [];
  lines.push(`const documents = {\n`);

  for (const { operations, ...rest } of sourcesWithOperations) {
    const originalString = rest.source.rawSDL!;
    const operation = operations[0];

    lines.push(`    ${JSON.stringify(originalString)}: graphql.${operation.initialName},\n`);
  }

  lines.push(`};\n`);

  return lines;
}

function getGqlOverloadChunk(sourcesWithOperations: Array<SourceWithOperations>) {
  const lines: Array<string> = [];

  // We intentionally don't use a <T extends keyof typeof documents> generic, because TS
  // would print very long `gql` function signatures (duplicating the source).
  for (const { operations, ...rest } of sourcesWithOperations) {
    const originalString = rest.source.rawSDL!;
    lines.push(
      `export function gql(source: ${JSON.stringify(originalString)}): (typeof documents)[${JSON.stringify(
        originalString
      )}];\n`
    );
  }

  return lines;
}
