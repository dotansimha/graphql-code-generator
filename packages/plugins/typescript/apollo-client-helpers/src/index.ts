import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { GraphQLSchema, isInterfaceType, isObjectType } from 'graphql';
import { extname } from 'path';
import { ApolloClientHelpersConfig } from './config';

export const plugin: PluginFunction<ApolloClientHelpersConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: ApolloClientHelpersConfig
) => {
  const results: Types.ComplexPluginOutput[] = [];

  results.push(generateTypePoliciesSignature(schema, config));

  return {
    prepend: results.reduce((prev, r) => [...prev, ...(r.prepend || [])], []),
    append: results.reduce((prev, r) => [...prev, ...(r.append || [])], []),
    content: results.map(r => r.content).join('\n'),
  };
};

function generateTypePoliciesSignature(
  schema: GraphQLSchema,
  config: ApolloClientHelpersConfig
): Types.ComplexPluginOutput {
  const typeMap = schema.getTypeMap();
  const perTypePolicies: string[] = [];
  const typedTypePolicies = Object.keys(typeMap).reduce((prev, typeName) => {
    const type = typeMap[typeName];

    if (!typeName.startsWith('__') && (isObjectType(type) || isInterfaceType(type))) {
      const fieldsNames = Object.keys(type.getFields()).filter(f => !f.startsWith('__'));
      const keySpecifierVarName = `${typeName}KeySpecifier`;
      const fieldPolicyVarName = `${typeName}FieldPolicy`;

      perTypePolicies.push(
        `export type ${keySpecifierVarName} = (${fieldsNames
          .map(f => `'${f}'`)
          .join(' | ')} | ${keySpecifierVarName})[];`
      );

      perTypePolicies.push(`export type ${fieldPolicyVarName} = {
${fieldsNames.map(fieldName => `\t${fieldName}?: FieldPolicy<any> | FieldReadFunction<any>`).join(',\n')}
};`);

      return {
        ...prev,
        [typeName]: `Omit<TypePolicy, "fields" | "keyFields"> & {
\t\tkeyFields${
          config.requireKeyFields ? '' : '?'
        }: false | ${keySpecifierVarName} | (() => undefined | ${keySpecifierVarName}),
\t\tfields?: ${fieldPolicyVarName},
\t}`,
      };
    }

    return prev;
  }, {} as Record<string, string>);

  const rootTypes = [
    schema.getQueryType()?.name,
    schema.getMutationType()?.name,
    schema.getSubscriptionType()?.name,
  ].filter(Boolean);

  const rootContent = `export type TypedTypePolicies = TypePolicies & {${Object.keys(typedTypePolicies)
    .map(typeName => {
      const nonOptional = config.requirePoliciesForAllTypes && !rootTypes.includes(typeName);

      return `\n\t${typeName}${nonOptional ? '' : '?'}: ${typedTypePolicies[typeName]}`;
    })
    .join(',')}\n};`;

  return {
    prepend: [
      `import ${
        config.useTypeImports ? 'type ' : ''
      }{ FieldPolicy, FieldReadFunction, TypePolicies, TypePolicy } from '@apollo/client/cache';`,
    ],
    content: [...perTypePolicies, rootContent].join('\n'),
  };
}

export const validate: PluginValidateFn<ApolloClientHelpersConfig> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts' && extname(outputFile) !== '.tsx') {
    throw new Error(`Plugin "apollo-client-helpers" requires extension to be ".ts" or ".tsx"!`);
  }
};
