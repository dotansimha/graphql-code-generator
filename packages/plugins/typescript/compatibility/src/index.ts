import { GraphQLSchema, concatAST, visit } from 'graphql';
import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { CompatibilityPluginVisitor } from './visitor';
import { CompatibilityPluginRawConfig } from './config';

const REACT_APOLLO_PLUGIN_NAME = 'typescript-react-apollo';

export const plugin: PluginFunction<CompatibilityPluginRawConfig> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: CompatibilityPluginRawConfig,
  additionalData
): Promise<string> => {
  const allAst = concatAST(documents.map(v => v.document));

  const reactApollo = ((additionalData || {}).allPlugins || []).find(
    p => Object.keys(p)[0] === REACT_APOLLO_PLUGIN_NAME
  );
  const visitor = new CompatibilityPluginVisitor(config, schema, {
    reactApollo: reactApollo
      ? {
          ...(config || {}),
          ...(reactApollo[REACT_APOLLO_PLUGIN_NAME] as object),
        }
      : null,
  });

  const visitorResult = visit(allAst, {
    leave: visitor as any,
  });

  const discriminateUnion = `type DiscriminateUnion<T, U> = T extends U ? T : never;\n`;
  const result: string = visitorResult.definitions.filter(a => a && typeof a === 'string').join('\n');

  return result.includes('DiscriminateUnion') ? [discriminateUnion, result].join('\n') : result;
};
