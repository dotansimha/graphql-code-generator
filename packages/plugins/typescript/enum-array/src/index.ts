import { GraphQLSchema, GraphQLEnumType } from 'graphql';
import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { EnumArrayPluginConfig } from './config';

function getEnumTypeMap(schema: GraphQLSchema): GraphQLEnumType[] {
  const typeMap = schema.getTypeMap();
  const result: GraphQLEnumType[] = [];
  for (const key in typeMap) {
    if (typeMap[key].astNode?.kind === 'EnumTypeDefinition') {
      result.push(typeMap[key] as GraphQLEnumType);
    }
  }
  return result;
}

function buildArrayDefinition(e: GraphQLEnumType): string {
  const upperName = e.name
    .replace(/[A-Z]/g, letter => `_${letter}`)
    .slice(1)
    .toUpperCase();
  const values = e
    .getValues()
    .map(v => `'${v.value}'`)
    .join(', ');
  return `export const ${upperName}: ${e.name}[] = [${values}];`;
}

function buildImportStatement(enums: GraphQLEnumType[], importFrom: string): string[] {
  const names: string[] = Object.values(enums).map(e => e.name);
  return [`import { ${names.join(', ')} } from "${importFrom}";`];
}

export const plugin: PluginFunction<EnumArrayPluginConfig> = async (
  schema: GraphQLSchema,
  _documents: Types.DocumentFile[],
  config: EnumArrayPluginConfig
): Promise<Types.PluginOutput> => {
  const importFrom: EnumArrayPluginConfig['importFrom'] = config.importFrom;
  const enums = getEnumTypeMap(schema);
  const content = enums.map(buildArrayDefinition).join('\n');

  const result: Types.PluginOutput = { content };
  if (importFrom) {
    result['prepend'] = buildImportStatement(enums, importFrom);
  }
  return result;
};

export default { plugin };
