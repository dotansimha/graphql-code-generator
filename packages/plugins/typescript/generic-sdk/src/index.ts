import { PluginFunction, PluginValidateFn, Types } from '@graphql-codegen/plugin-helpers';
import { LoadedFragment, RawClientSideBasePluginConfig } from '@graphql-codegen/visitor-plugin-common';
import { concatAST, FragmentDefinitionNode, GraphQLSchema, Kind, visit } from 'graphql';
import { extname } from 'path';
import { RawGenericSdkPluginConfig } from './config';
import { GenericSdkVisitor } from './visitor';

export const plugin: PluginFunction<RawGenericSdkPluginConfig> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: RawGenericSdkPluginConfig
) => {
  const allAst = concatAST(
    documents.reduce((prev, v) => {
      return [...prev, v.document];
    }, [])
  );
  const allFragments: LoadedFragment[] = [
    ...(allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION) as FragmentDefinitionNode[]).map(
      fragmentDef => ({
        node: fragmentDef,
        name: fragmentDef.name.value,
        onType: fragmentDef.typeCondition.name.value,
        isExternal: false,
      })
    ),
    ...(config.externalFragments || []),
  ];
  const visitor = new GenericSdkVisitor(schema, allFragments, config);
  const visitorResult = visit(allAst, { leave: visitor });

  return {
    prepend: visitor.getImports(),
    content: [
      visitor.fragments,
      ...visitorResult.definitions.filter(t => typeof t === 'string'),
      visitor.sdkContent,
    ].join('\n'),
  };
};

export const validate: PluginValidateFn<any> = async (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: RawClientSideBasePluginConfig,
  outputFile: string
) => {
  if (extname(outputFile) !== '.ts') {
    throw new Error(`Plugin "typescript-generic-sdk" requires extension to be ".ts"!`);
  }
};

export { GenericSdkVisitor };
