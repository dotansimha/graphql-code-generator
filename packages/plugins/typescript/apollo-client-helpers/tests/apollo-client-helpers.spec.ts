import { mergeOutputs } from '@graphql-codegen/plugin-helpers';
import { buildSchema } from 'graphql';
import { plugin } from '../src/index';

describe('apollo-client-helpers', () => {
  it('Should output typePolicies object correctly', async () => {
    const schema = buildSchema(/* GraphQL */ `
      type Query {
        user: User!
      }
      type User {
        id: ID!
        name: String!
      }
    `);
    const result = mergeOutputs([await plugin(schema, [], {})]);
    expect(result).toMatchSnapshot();
  });

  it('Should output typePolicies object with requireKeyFields: true', async () => {
    const schema = buildSchema(/* GraphQL */ `
      type Query {
        user: User!
      }
      type User {
        id: ID!
        name: String!
      }
    `);
    const result = mergeOutputs([
      await plugin(schema, [], {
        requireKeyFields: true,
      }),
    ]);
    expect(result).toContain(`keyFields:`);
    expect(result).not.toContain(`keyFields?:`);
    expect(result).toMatchSnapshot();
  });

  it('Should output typePolicies object with requirePoliciesForAllTypes: true', async () => {
    const schema = buildSchema(/* GraphQL */ `
      type Query {
        user: User!
      }
      type User {
        id: ID!
        name: String!
      }
    `);
    const result = mergeOutputs([
      await plugin(schema, [], {
        requirePoliciesForAllTypes: true,
      }),
    ]);
    expect(result).toContain(`User:`);
    expect(result).toContain(`Query?:`);
    expect(result).toMatchSnapshot();
  });
});
