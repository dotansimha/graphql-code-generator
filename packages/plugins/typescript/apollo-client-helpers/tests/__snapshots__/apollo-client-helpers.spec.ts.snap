// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apollo-client-helpers Should output typePolicies object correctly 1`] = `
"import { FieldPolicy, FieldReadFunction, TypePolicies, TypePolicy } from '@apollo/client/cache';
export type QueryKeySpecifier = ('user' | QueryKeySpecifier)[];
export type QueryFieldPolicy = {
	user?: FieldPolicy<any> | FieldReadFunction<any>
};
export type UserKeySpecifier = ('id' | 'name' | UserKeySpecifier)[];
export type UserFieldPolicy = {
	id?: FieldPolicy<any> | FieldReadFunction<any>,
	name?: FieldPolicy<any> | FieldReadFunction<any>
};
export type TypedTypePolicies = TypePolicies & {
	Query?: Omit<TypePolicy, \\"fields\\" | \\"keyFields\\"> & {
		keyFields?: false | QueryKeySpecifier | (() => undefined | QueryKeySpecifier),
		fields?: QueryFieldPolicy,
	},
	User?: Omit<TypePolicy, \\"fields\\" | \\"keyFields\\"> & {
		keyFields?: false | UserKeySpecifier | (() => undefined | UserKeySpecifier),
		fields?: UserFieldPolicy,
	}
};"
`;

exports[`apollo-client-helpers Should output typePolicies object with requireKeyFields: true 1`] = `
"import { FieldPolicy, FieldReadFunction, TypePolicies, TypePolicy } from '@apollo/client/cache';
export type QueryKeySpecifier = ('user' | QueryKeySpecifier)[];
export type QueryFieldPolicy = {
	user?: FieldPolicy<any> | FieldReadFunction<any>
};
export type UserKeySpecifier = ('id' | 'name' | UserKeySpecifier)[];
export type UserFieldPolicy = {
	id?: FieldPolicy<any> | FieldReadFunction<any>,
	name?: FieldPolicy<any> | FieldReadFunction<any>
};
export type TypedTypePolicies = TypePolicies & {
	Query?: Omit<TypePolicy, \\"fields\\" | \\"keyFields\\"> & {
		keyFields: false | QueryKeySpecifier | (() => undefined | QueryKeySpecifier),
		fields?: QueryFieldPolicy,
	},
	User?: Omit<TypePolicy, \\"fields\\" | \\"keyFields\\"> & {
		keyFields: false | UserKeySpecifier | (() => undefined | UserKeySpecifier),
		fields?: UserFieldPolicy,
	}
};"
`;

exports[`apollo-client-helpers Should output typePolicies object with requirePoliciesForAllTypes: true 1`] = `
"import { FieldPolicy, FieldReadFunction, TypePolicies, TypePolicy } from '@apollo/client/cache';
export type QueryKeySpecifier = ('user' | QueryKeySpecifier)[];
export type QueryFieldPolicy = {
	user?: FieldPolicy<any> | FieldReadFunction<any>
};
export type UserKeySpecifier = ('id' | 'name' | UserKeySpecifier)[];
export type UserFieldPolicy = {
	id?: FieldPolicy<any> | FieldReadFunction<any>,
	name?: FieldPolicy<any> | FieldReadFunction<any>
};
export type TypedTypePolicies = TypePolicies & {
	Query?: Omit<TypePolicy, \\"fields\\" | \\"keyFields\\"> & {
		keyFields?: false | QueryKeySpecifier | (() => undefined | QueryKeySpecifier),
		fields?: QueryFieldPolicy,
	},
	User: Omit<TypePolicy, \\"fields\\" | \\"keyFields\\"> & {
		keyFields?: false | UserKeySpecifier | (() => undefined | UserKeySpecifier),
		fields?: UserFieldPolicy,
	}
};"
`;
